src/category_theory/limits/shapes/strict_initial.lean

```diff
@@ -0,0 +1,161 @@
+initial objects exist, which turns out to be a more useful notion to formalise.
+
+If the binary product of `X` with a strict initial object exists, it is also initial.
+
+To show a category `C` with an initial object has strict initial objects, the most convenient way
+is to show any morphism to the (chosen) initial object is an isomorphism and use
+`has_strict_initial_object_of_initial_is_strict`.
+
+The dual notion (strict terminal objects) occurs much less frequently in practice so is ignored.
+
+## TODO
+
+* Construct examples of this: `Type*`, `Top`, `Groupoid`, simplicial types, posets.
+* Construct the bottom element of the subobject lattice given strict initials.
+* Show cartesian closed categories have strict initials
+
+## References
+* https://ncatlab.org/nlab/show/strict+initial+object
+-/
+
+
+universes v u
+
+namespace category_theory
+namespace limits
+open category
+
+variables (C : Type u) [category.{v} C]
+
+/--
+We say `C` has strict initial objects if every initial object is strict, ie given any morphism
+`f : A ⟶ I` where `I` is initial, then `f` is an isomorphism.
+
+Strictly speaking, this says that *any* initial object must be strict, rather than that strict
+initial objects exist.
+-/
+class has_strict_initial_object : Prop :=
+(out : ∀ {I A : C} (f : A ⟶ I), is_initial I → is_iso f)
+
+variables {C}
+
+section
+variables [has_strict_initial_object C] {I : C}
+
+lemma is_initial.is_iso_to (hI : is_initial I) {A : C} (f : A ⟶ I) :
+  is_iso f :=
+has_strict_initial_object.out f hI
+
+lemma is_initial.subsingleton_to (hI : is_initial I) {A : C} (f g : A ⟶ I) :
+  f = g :=
+begin
+  haveI := hI.is_iso_to f,
+  haveI := hI.is_iso_to g,
+  exact eq_of_inv_eq_inv (hI.hom_ext (inv f) (inv g)),
+end
+
+lemma is_initial.mono_from (hI : is_initial I) {A : C} (f : I ⟶ A) :
+  mono f :=
+{ right_cancellation := λ B g h i, hI.subsingleton_to _ _ }
+
+lemma is_initial.to_mono (hI : is_initial I) {A : C} :
+  mono (hI.to A) :=
+hI.mono_from _
+
+/-- If `I` is initial, then `X ⨯ I` is isomorphic to it. -/
+@[simps hom]
+noncomputable def mul_is_initial (X : C) [has_binary_product X I] (hI : is_initial I) :
+  X ⨯ I ≅ I :=
+@@as_iso _ prod.snd (hI.is_iso_to _)
+
+@[simp] lemma mul_is_initial_inv (X : C) [has_binary_product X I] (hI : is_initial I) :
+  (mul_is_initial X hI).inv = hI.to _ :=
+hI.hom_ext _ _
+
+/-- If `I` is initial, then `I ⨯ X` is isomorphic to it. -/
+@[simps hom]
+noncomputable def is_initial_mul (X : C) [has_binary_product I X] (hI : is_initial I) :
+  I ⨯ X ≅ I :=
+@@as_iso _ prod.fst (hI.is_iso_to _)
+
+@[simp] lemma is_initial_mul_inv (X : C) [has_binary_product I X] (hI : is_initial I) :
+  (is_initial_mul X hI).inv = hI.to _ :=
+hI.hom_ext _ _
+
+variable [has_initial C]
+
+instance initial_is_iso_to {A : C} (f : A ⟶ ⊥_ C) : is_iso f :=
+initial_is_initial.is_iso_to _
+
+lemma initial_subsingleton_to {A : C} (f g : A ⟶ ⊥_ C) :
+  f = g :=
```

Comment:
Is there a reason why you didn't go with ` instance initial_subsingleton_to {A : C} : subsingleton (A ⟶ ⊥_ C)`?

Reply:
No good reason, I've amended! I kept the old versions too since I think they're sometimes useful in term mode, and changed the names to `_hom_ext` instead.